tests/gstreamer_mqtt/unittest_ntp_util_mock.cc

```diff
@@ -0,0 +1,317 @@
+                  socklen_t addrlen));
+  MOCK_METHOD (ssize_t, write, (int fd, const void *buf, size_t count));
+  MOCK_METHOD (ssize_t, read, (int fd, void *buf, size_t count));
+  MOCK_METHOD (uint32_t, _convert_to_host_byte_order, (uint32_t netlong));
+};
+NtpUtilMock *mockInstance = nullptr;
+
+/**
+ * @brief Mocking function for gethostbyname
+ */
+struct hostent *gethostbyname (const char *name)
+{
+  return mockInstance->gethostbyname (name);
+}
+
+/**
+ * @brief Mocking function for gethostbyname
+ */
+int connect (int sockfd, const struct sockaddr *addr,
+                  socklen_t addrlen)
+{
+  return mockInstance->connect (sockfd, addr, addrlen);
+}
+
+/** @note To avoid redundant declaration in the test */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wredundant-decls"
+/**
+ * @brief Mocking function for write.
+ */
+ssize_t write (int fd, const void *buf, size_t count)
+{
+  return mockInstance->write (fd, buf, count);
+}
+
+/**
+ * @brief Mocking function for read.
+ */
+ssize_t read (int fd, void *buf, size_t count)
+{
+  return mockInstance->read (fd, buf, count);
+}
+#pragma GCC diagnostic pop
+
+/**
+ * @brief Mocking function for _convert_to_host_byte_order.
+ */
+uint32_t _convert_to_host_byte_order (uint32_t netlong)
+{
+  return mockInstance->_convert_to_host_byte_order (netlong);
+}
+
+/**
+ * @brief Initialize hostent.
+ */
+static void _hostent_init (struct hostent *host)
+{
+  host->h_name = g_strdup ("github.com");
+  host->h_aliases = (char **) calloc (1, sizeof (char *));
+  host->h_aliases[0] = g_strdup ("www.github.com");
+  host->h_addrtype = AF_INET;
+  host->h_length = 4;
+  host->h_addr_list = (char **) calloc (1, sizeof (char *));
+  host->h_addr_list[0] = g_strdup ("52.78.231.108");
+}
+
+/**
+ * @brief Free hostent.
+ */
+static void _hostent_free (struct hostent *host)
+{
+  g_free (host->h_name);
+  g_free (host->h_aliases[0]);
+  g_free (host->h_aliases);
+  g_free (host->h_addr_list[0]);
+  g_free (host->h_addr_list);
+}
+
+/**
+ * @brief Test for ntp util to get epoch.
+ */
+TEST (NtpUtilMockGetEpoch, Normal_p)
+{
+  int64_t ret;
+  const char *hnames[] = {"temp"};
+  struct hostent temp_hostent;
+  uint16_t ports[] = {8080U};
+
+  mockInstance = new NtpUtilMock ();
+  _hostent_init (&temp_hostent);
+
+  EXPECT_CALL (*mockInstance, gethostbyname(_))
+      .Times(1).WillOnce(Return((struct hostent *)&temp_hostent));
+  EXPECT_CALL (*mockInstance, connect(_, _, _))
+      .Times(1).WillOnce(Return(0));
+  EXPECT_CALL (*mockInstance, write(_, _, _))
+      .Times(1).WillOnce(Return(0));
+  EXPECT_CALL (*mockInstance, read(_, _, _))
+      .Times(1).WillOnce(Return(0));
+  EXPECT_CALL (*mockInstance, _convert_to_host_byte_order(_))
+      .Times(2)
+      .WillOnce(Return(NTPUTIL_TIMESTAMP_DELTA + 1ULL))
+      .WillOnce(Return(1ULL));
+
+  ret = ntputil_get_epoch (1, (char **)hnames, ports);
+
+  EXPECT_GE (ret, 0);
+
+  _hostent_free (&temp_hostent);
+  delete mockInstance;
+}
```

Comment:
After that, we can run each test case without calling `_hostent_init()` and `_hostent_free() as below.
```cpp
TEST_F (NtpUtilMockTest, Normal_p)
{
  ...
  mockInstance = new NtpUtilMock ();

  EXPECT_CALL (*mockInstance, gethostbyname(_))
      .Times(1).WillOnce(Return((struct hostent *)&temp_hostent));
  ...

  ret = ntputil_get_epoch (1, (char **)hnames, ports);

  EXPECT_GE (ret, 0);
  ...
}
```



In my opinion, it is easy to maintain since the test framework automatically calls both SetUp() and TearDown() methods.

Reply:
Thanks! Applied.